go/vt/vtadmin/cluster/cluster.go

```diff
@@ -373,50 +374,269 @@ func (c *Cluster) getTablets(ctx context.Context) ([]*vtadminpb.Tablet, error) {
+			return tablet.Tablet.Keyspace == keyspace && tablet.State == vtadminpb.Tablet_SERVING
+		}, opts.Tablets, len(opts.Tablets))
+
+		if len(keyspaceTablets) == 0 {
+			// consider how to include info about the tablets we looked at, but
+			// that's also potentially a very long list .... maybe we should
+			// just log it (yes, do that).
+			return nil, fmt.Errorf("%w for keyspace %s", errors.ErrNoServingTablet, keyspace)
+		}
+
+		randomServingTablet := keyspaceTablets[rand.Intn(len(keyspaceTablets))]
+		tabletsToQuery = append(tabletsToQuery, randomServingTablet)
 	}
 
-	return &vtadminpb.Schema{
-		Cluster:          c.ToProto(),
-		Keyspace:         tablet.Tablet.Keyspace,
-		TableDefinitions: schema.Schema.TableDefinitions,
-	}, nil
+	var (
+		m      sync.Mutex
+		wg     sync.WaitGroup
+		rec    concurrency.AllErrorRecorder
+		schema = &vtadminpb.Schema{
+			Cluster:    c.ToProto(),
+			Keyspace:   keyspace,
+			TableSizes: map[string]*vtadminpb.Schema_TableSize{},
+		}
+		// Instead of starting at false, we start with whatever the base request
+		// specified. If we have exactly one tablet to query (i.e. we're not
+		// doing multi-shard aggregation), it's possible the request was to
+		// literally just get the table sizes; we shouldn't assume. If we have
+		// more than one tablet to query, then we are doing size aggregation,
+		// and we'll flip this to true after spawning the first GetSchema rpc.
+		sizesOnly = opts.BaseRequest.TableSizesOnly
+	)
+
+	for _, tablet := range tabletsToQuery {
+		wg.Add(1)
+
+		go func(tablet *vtadminpb.Tablet, sizesOnly bool) {
+			defer wg.Done()
+
+			span, ctx := trace.NewSpan(ctx, "Vtctld.GetSchema")
+			defer span.Finish()
+
+			req := *opts.BaseRequest
+			req.TableSizesOnly = sizesOnly
+			req.TabletAlias = tablet.Tablet.Alias
+
+			AnnotateSpan(c, span)
+			annotateGetSchemaRequest(&req, span)
+			span.Annotate("keyspace", keyspace)
+			span.Annotate("shard", tablet.Tablet.Shard)
+
+			resp, err := c.Vtctld.GetSchema(ctx, &req)
+			if err != nil {
+				err = fmt.Errorf("GetSchema(cluster = %s, keyspace = %s, tablet = %s) failed: %w", c.ID, keyspace, tablet.Tablet.Alias, err)
+				rec.RecordError(err)
+
+				return
+			}
+
+			if resp == nil || resp.Schema == nil {
+				return
+			}
+
+			m.Lock()
+			defer m.Unlock()
+
+			if !sizesOnly {
+				schema.TableDefinitions = resp.Schema.TableDefinitions
+			}
+
+			if !opts.SizeOpts.AggregateSizes {
+				return
+			}
+
+			for _, td := range resp.Schema.TableDefinitions {
+				tableSize, ok := schema.TableSizes[td.Name]
+				if !ok {
+					tableSize = &vtadminpb.Schema_TableSize{
+						ByShard: map[string]*vtadminpb.Schema_ShardTableSize{},
+					}
+					schema.TableSizes[td.Name] = tableSize
+				}
+
+				if _, ok = tableSize.ByShard[tablet.Tablet.Shard]; ok {
+					// We managed to query for the same shard twice, that's ...
+					// weird. but do we care? maybe just log? idk!
```

Comment:
lol I do truly love this comment, but it is somewhat out of character with the rest so I wanted to point it out _just in case_

Reply:
Maybe we should error? That seems like it shouldn't happen, so if it does happen we better investigate??